File: src/main/java/com/jcloisterzone/game/phase/GamblersLuckDicePhase.java
package com.jcloisterzone.game.phase;

import com.jcloisterzone.board.Tile;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.event.DiceSixRollEvent;
import com.jcloisterzone.event.PlayEvent.PlayEventMeta;
import com.jcloisterzone.feature.City;
import com.jcloisterzone.feature.GamblersLuckShield;
import com.jcloisterzone.feature.Road;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.modifier.FeatureModifier;
import com.jcloisterzone.game.Rule;
import com.jcloisterzone.game.capability.FamiliesCapability;
import com.jcloisterzone.game.capability.GamblersLuckCapability;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import com.jcloisterzone.random.RandomGenerator;

import io.vavr.Tuple2;
import io.vavr.Tuple3;
import io.vavr.collection.HashMap;
import io.vavr.collection.HashSet;
import io.vavr.collection.Map;
import io.vavr.collection.LinkedHashMap;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;
import io.vavr.collection.TreeMap;

import java.util.function.Function;

public class GamblersLuckDicePhase extends Phase {

	public static final Map<Integer, Tuple3<Integer, String, GamblersLuckCapability.GamblersLuckShieldToken>> DICE_REGULAR = TreeMap.of(
		1, new Tuple3(1, "grey", GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_1),
		2, new Tuple3(2, "grey" /*both*/, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_2),
		3, new Tuple3(3, "grey", GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_3),
		4, new Tuple3(null, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_X),
		5, new Tuple3(0, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_0),
		6, new Tuple3(0, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_0)
	);
	
	public static final Map<Integer, Tuple3<Integer, String, GamblersLuckCapability.GamblersLuckShieldToken>> DICE_MAYOR = TreeMap.of(
			1, new Tuple3(1, "grey", GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_1),
			2, new Tuple3(2, "grey", GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_2),
			3, new Tuple3(null, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_X),
			4, new Tuple3(null, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_X),
			5, new Tuple3(0, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_0),
			6, new Tuple3(0, null, GamblersLuckCapability.GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_0)
		);

	public GamblersLuckDicePhase(RandomGenerator random, Phase defaultNext) {
        super(random, defaultNext);
    }

    @Override
    public StepResult enter(GameState state) {
        Position pos = state.getLastPlaced().getPosition();

        GamblersLuckCapability cap = state.getCapabilities().get(GamblersLuckCapability.class);
        
        Stream<Tuple2<FeaturePointer,GamblersLuckShield>> shields = cap.getPlacedTileGamblersLuckShields(state);
        
        if (shields.length() == 0) {
        	return next(state);
        }

        Map<FeaturePointer, Tuple2<GamblersLuckCapability.GamblersLuckShieldToken,Integer>> model = cap.getModel(state);

        Map<Integer, Tuple3<Integer, String, GamblersLuckCapability.GamblersLuckShieldToken>> dice = DICE_REGULAR;
        if (state.getElements().get("mayor").getOrNull() != null) {
        	dice = DICE_MAYOR;
        }
        
    	for(Tuple2<FeaturePointer,GamblersLuckShield> shield: shields) {
        	int diceValue = state.getPhase().getRandom().getNextInt(6)+1;
        	int randomRotation = state.getPhase().getRandom().getNextInt(8)-5;
        	model = model.put(shield._1, new Tuple2(dice.get(diceValue).get()._3,randomRotation));
			state = state.appendEvent(new DiceSixRollEvent(
	                PlayEventMeta.createWithActivePlayer(state), HashSet.of(pos), diceValue, String.format("%s.%s", "gamblers-luck-dice", dice.get(diceValue).get()._3)));
        }
        
        state = cap.setModel(state, model);

		Map<FeaturePointer,City> cities = state
				.getPlacedTile(pos)
				.getTile()
				.getInitialFeatures()
				.filterValues(f -> f instanceof City)
				.mapValues(f -> (City) f);

		for(Tuple2<FeaturePointer,City> city: cities) {
			Stream<FeaturePointer> initialShields = city._2.getModifier(state, City.GAMBLERS_LUCK_SHIELDS, Stream.empty() );
			Stream<FeaturePointer> placedShields = Stream.empty();
			int newShields = city._2.getModifier(state, City.PENNANTS, 0);
			boolean eliminatedPennats = false;
			String family = null;
			for(FeaturePointer initialShield: initialShields) {
				FeaturePointer placedShield = initialShield.setPosition(pos);
				placedShields = placedShields.append(placedShield);
				Tuple2<GamblersLuckCapability.GamblersLuckShieldToken,Integer> currentToken = model.get(placedShield).getOrNull();
				Tuple3<Integer, String, GamblersLuckCapability.GamblersLuckShieldToken> diceValue = findByToken(dice, currentToken._1);
				if (diceValue._1 == null) {
					eliminatedPennats = true;
				} else if (diceValue._1 > 0) {
					newShields += diceValue._1;
					if (family == null || !family.equals("both")) {
						family = diceValue._2;
					}
				}
			}
			Map<FeatureModifier<?>, Object> modifiers = city._2.getModifiers();
			if (eliminatedPennats) {
				modifiers = modifiers.put(City.ELIMINATED_PENNANTS, eliminatedPennats);
			}
			if (newShields>0) {
				modifiers = modifiers.put(City.PENNANTS, newShields);
				modifiers = modifiers.put(FamiliesCapability.FAMILY, family);
			}
			modifiers = modifiers.replace(City.GAMBLERS_LUCK_SHIELDS,initialShields,placedShields);

			state = updatedInitialFeatures(state, pos, city, modifiers);
		}

        
        return next(state);
    }
    
    public static Tuple3<Integer, String, GamblersLuckCapability.GamblersLuckShieldToken>
    	findByToken(
    			Map<Integer, Tuple3<Integer, String, GamblersLuckCapability.GamblersLuckShieldToken>> dice,
    			GamblersLuckCapability.GamblersLuckShieldToken token) {

        return dice.values()
            .filter(t -> t._3 == token)
            .getOrNull();
    }
    
    private GameState updatedInitialFeatures(GameState state, Position pos,
            Tuple2<FeaturePointer, City> city,
            Map<FeatureModifier<?>, Object> modifiers) {

        // get original tile
        Tile tile = state.getPlacedTile(pos).getTile();
        Map<FeaturePointer, Feature> features = tile.getInitialFeatures();

        final FeaturePointer cityKey = city._1;
        final City cityValue = city._2;
        final Map<FeatureModifier<?>, Object> modifiersFinal = modifiers;

        // update features in the tile
        Tile updatedTile = tile.setInitialFeatures(
            features.put(
                cityKey,
                features.get(cityKey)
                        .map(f -> ((City) f).setModifiers(modifiersFinal))
                        .getOrElse(() -> cityValue.setModifiers(modifiersFinal))
            )
        );
        LinkedHashMap<Position, PlacedTile> placedTiles = state.getPlacedTiles();

        PlacedTile updatedPlacedTile = placedTiles.get(pos)
            .map(pt -> pt.setTile(updatedTile))
            .getOrElseThrow(() -> new IllegalArgumentException("No tile at position " + pos));

        placedTiles = placedTiles.put(pos, updatedPlacedTile);
        
        state = state.setPlacedTiles(placedTiles);
        
        // Initial Features are updated, now we are going to recalculate city modifier by them
        
        FeaturePointer cityOnBoard = city._1.setPosition(pos).rotateCW(state.getPlacedTile(pos).getRotation());
        
        List<FeaturePointer> places = state.getFeature(cityOnBoard).getPlaces();

        final GameState capturedState = state;  // Capture in final variable
        List<City> initialFeatures = places
            .map(fp -> {
            	PlacedTile pt = capturedState.getPlacedTile(fp.getPosition());
                return (City) pt.getInitialFeaturePartOf(fp.getLocation())._2
                             .placeOnBoard(fp.getPosition(), pt.getRotation());
            });

        Map<FeatureModifier<?>, Object> EMPTY = HashMap.empty();
        Map<FeatureModifier<?>, Object> updatedModifiers = initialFeatures.foldLeft(EMPTY, (res, _city) -> _city.mergeModifiers(res));

        City updateCity = city._2.setModifiers(updatedModifiers);
        
        state = updateCityComplex(state, cityOnBoard, updatedModifiers);
      
        return state;
    }
    
    /**
     * Scenario: Update a city with multiple changes to initialFeatures
     * (e.g., add modifiers, close edges, update state)
     */
    public GameState updateCityComplex(GameState state, FeaturePointer cityPointer, Map<FeatureModifier<?>, Object> modifiers) {
        // Step 1: Get the original city
        City originalCity = (City) state.getFeature(cityPointer);
        
        // Step 2: Apply multiple transformations
        City updatedCity = originalCity
            .setModifiers(modifiers);
        
        // Step 3: Build the update map for all city pointers
        // This ensures every FeaturePointer pointing to this city
        // in the featureMap now points to the updated version
        Map<FeaturePointer, Feature> featureMapUpdate = originalCity.getPlaces()
            .foldLeft(
                HashMap.empty(),
                (map, fp) -> map.put(fp, updatedCity)
            );
        
        // Step 4: Apply update
        return state.updateFeatureMap(featureMapUpdate);
    }
}

File: src/main/ts/com/jcloisterzone/game/phase/GamblersLuckDicePhase.ts
import { HashSet } from "../../../../io/vavr/Set.js";
import { Stream, List } from "../../../../io/vavr/SeqTypes.js";
import { Tuple2 } from "../../../../io/vavr/Tuple.js";
import { HashMap, type Map as VMap } from "../../../../io/vavr/Map.js";
import type { ClassToken } from "../../../../lang/Class.js";
import type { Position } from "../../board/Position.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import { DiceSixRollEvent } from "../../event/DiceSixRollEvent.js";
import { PlayEventMeta } from "../../event/PlayEvent.js";
import { City } from "../../feature/City.js";
import type { Feature } from "../../feature/Feature.js";
import type { FeatureModifier } from "../../feature/modifier/FeatureModifier.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { GamblersLuckCapability } from "../capability/GamblersLuckCapability.js";
import { FamiliesCapability } from "../capability/FamiliesCapability.js";
import type { GameState } from "../state/GameState.js";
import { Phase } from "./Phase.js";
import type { StepResult } from "./StepResult.js";

type ShieldToken = GamblersLuckCapability.GamblersLuckShieldToken;
type DiceRow = [number | null, string | null, ShieldToken];
type ModMap = VMap<FeatureModifier<unknown>, unknown>;

const T = GamblersLuckCapability.GamblersLuckShieldToken;

/** die value → [pennants gained (null = eliminates pennants), family colour, token]. */
const DICE_REGULAR = new globalThis.Map<number, DiceRow>([
  [1, [1, "grey", T.GAMBLERSLUCKSHIELD_1]],
  [2, [2, "grey" /* both */, T.GAMBLERSLUCKSHIELD_2]],
  [3, [3, "grey", T.GAMBLERSLUCKSHIELD_3]],
  [4, [null, null, T.GAMBLERSLUCKSHIELD_X]],
  [5, [0, null, T.GAMBLERSLUCKSHIELD_0]],
  [6, [0, null, T.GAMBLERSLUCKSHIELD_0]],
]);
const DICE_MAYOR = new globalThis.Map<number, DiceRow>([
  [1, [1, "grey", T.GAMBLERSLUCKSHIELD_1]],
  [2, [2, "grey", T.GAMBLERSLUCKSHIELD_2]],
  [3, [null, null, T.GAMBLERSLUCKSHIELD_X]],
  [4, [null, null, T.GAMBLERSLUCKSHIELD_X]],
  [5, [0, null, T.GAMBLERSLUCKSHIELD_0]],
  [6, [0, null, T.GAMBLERSLUCKSHIELD_0]],
]);

function findByToken(dice: globalThis.Map<number, DiceRow>, token: ShieldToken): DiceRow {
  for (const row of dice.values()) {
    if (row[2] === token) return row;
  }
  throw new Error("Unknown shield token");
}

const CAP_CLS = GamblersLuckCapability as unknown as ClassToken<GamblersLuckCapability>;

/** After the action commit, rolls a d6 for every gamblers-luck shield on the placed
 *  tile: the result becomes a shield token (+ random extra rotation for the client)
 *  and rewrites the tile's city pennants / family modifiers accordingly. */
export class GamblersLuckDicePhase extends Phase {
  static readonly simpleName = "GamblersLuckDicePhase";

  constructor(random: RandomGenerator, defaultNext: Phase | null) {
    super(random, defaultNext);
  }

  enter(state: GameState): StepResult {
    const pos = state.getLastPlaced()!.getPosition();
    const cap = state.getCapabilities().get(CAP_CLS) as GamblersLuckCapability;

    const shields = cap.getPlacedTileGamblersLuckShields(state);
    if (shields.length() === 0) {
      return this.next(state);
    }

    let model = cap.getModel(state) as VMap<FeaturePointer, Tuple2<ShieldToken, number>>;
    const dice =
      state.getElements().get("mayor").getOrNull() !== null &&
      state.getElements().get("mayor").getOrNull() !== undefined
        ? DICE_MAYOR
        : DICE_REGULAR;

    for (const shield of shields) {
      const diceValue = this.getRandom().getNextInt(6) + 1;
      const randomRotation = this.getRandom().getNextInt(8) - 5;
      model = model.put(shield._1, new Tuple2(dice.get(diceValue)![2], randomRotation));
      state = state.appendEvent(
        new DiceSixRollEvent(
          PlayEventMeta.createWithActivePlayer(state),
          HashSet.of(pos),
          diceValue,
          `gamblers-luck-dice.${dice.get(diceValue)![2].name()}`,
        ),
      );
    }
    state = cap.setModel(state, model);

    for (const cityEntry of state.getPlacedTile(pos)!.getTile().getInitialFeatures()) {
      if (!(cityEntry._2 instanceof City)) continue;
      const city = cityEntry._2;
      const initialShields = city.getModifier(
        state,
        City.GAMBLERS_LUCK_SHIELDS,
        Stream.empty<FeaturePointer>(),
      );
      let placedShields: Stream<FeaturePointer> = Stream.empty<FeaturePointer>();
      let newShields = city.getModifier(state, City.PENNANTS, 0);
      let eliminatedPennants = false;
      let family: string | null = null;
      for (const initialShield of initialShields) {
        const placedShield = initialShield.setPosition(pos);
        placedShields = placedShields.append(placedShield) as Stream<FeaturePointer>;
        const currentToken = model.get(placedShield).getOrNull();
        const diceRow = findByToken(dice, currentToken!._1);
        if (diceRow[0] === null) {
          eliminatedPennants = true;
        } else if (diceRow[0] > 0) {
          newShields += diceRow[0];
          if (family === null || family !== "both") {
            family = diceRow[1];
          }
        }
      }
      let modifiers: ModMap = city.getModifiers();
      if (eliminatedPennants) {
        modifiers = modifiers.put(City.ELIMINATED_PENNANTS as unknown as FeatureModifier<unknown>, eliminatedPennants);
      }
      if (newShields > 0) {
        modifiers = modifiers.put(City.PENNANTS as unknown as FeatureModifier<unknown>, newShields);
        modifiers = modifiers.put(FamiliesCapability.FAMILY as unknown as FeatureModifier<unknown>, family);
      }
      // vavr Map.replace(key, oldValue, newValue): only replaces an existing mapping
      const key = City.GAMBLERS_LUCK_SHIELDS as unknown as FeatureModifier<unknown>;
      if (modifiers.containsKey(key)) {
        modifiers = modifiers.put(key, placedShields);
      }

      state = this.updatedInitialFeatures(state, pos, new Tuple2(cityEntry._1, city), modifiers);
    }

    return this.next(state);
  }

  private updatedInitialFeatures(
    state: GameState,
    pos: Position,
    city: Tuple2<FeaturePointer, City>,
    modifiers: ModMap,
  ): GameState {
    // update the city's modifiers inside the placed tile's initial features
    const tile = state.getPlacedTile(pos)!.getTile();
    const features = tile.getInitialFeatures();
    const updatedTile = tile.setInitialFeatures(
      features.put(
        city._1,
        features
          .get(city._1)
          .map((f) => (f as City).setModifiers(modifiers) as unknown as Feature)
          .getOrElse(() => city._2.setModifiers(modifiers) as unknown as Feature),
      ),
    );
    let placedTiles = state.getPlacedTiles();
    const updatedPlacedTile = placedTiles
      .get(pos)
      .map((pt) => pt.setTile(updatedTile))
      .getOrElseThrow(() => new Error("No tile at position " + pos));
    placedTiles = placedTiles.put(pos, updatedPlacedTile) as typeof placedTiles;
    state = state.setPlacedTiles(placedTiles);

    // recompute the merged on-board city's modifiers from its tiles' initial features
    const cityOnBoard = city._1.setPosition(pos).rotateCW(state.getPlacedTile(pos)!.getRotation());
    if (state.getFeature(cityOnBoard) === null) {
      // the city was already transformed to a castle — skip
      return state;
    }

    const places = state.getFeature(cityOnBoard)!.getPlaces();
    const initialFeatures = places.map((fp) => {
      const pt = state.getPlacedTile(fp.getPosition())!;
      return pt.getInitialFeaturePartOf(fp.getLocation()!)!._2.placeOnBoard(
        fp.getPosition(),
        pt.getRotation(),
      ) as City;
    }) as List<City>;

    let updatedModifiers: ModMap = HashMap.empty<FeatureModifier<unknown>, unknown>();
    for (const c of initialFeatures) {
      updatedModifiers = c.mergeModifiers(updatedModifiers);
    }

    return this.updateCityComplex(state, cityOnBoard, updatedModifiers);
  }

  /** Re-point every FeaturePointer of the merged city at a modifier-updated copy. */
  private updateCityComplex(
    state: GameState,
    cityPointer: FeaturePointer,
    modifiers: ModMap,
  ): GameState {
    const originalCity = state.getFeature(cityPointer) as City;
    const updatedCity = originalCity.setModifiers(modifiers);
    let featureMapUpdate: VMap<FeaturePointer, Feature> = HashMap.empty<FeaturePointer, Feature>();
    for (const fp of originalCity.getPlaces()) {
      featureMapUpdate = featureMapUpdate.put(fp, updatedCity as unknown as Feature);
    }
    return state.updateFeatureMap(featureMapUpdate);
  }
}
